feat(Timeseries): Respect time grain

gerbermichi opened the pull request:
## **User description**
### SUMMARY
Respect time grain in timeseries and mixed timeseries chart.

### BEFORE/AFTER SCREENSHOTS OR ANIMATED GIF

#### BEFORE
<img width="2394" alt="image" src="https://github.com/user-attachments/assets/31f89bdd-cfd9-48c3-8841-5d24946e4a65" />

#### AFTER
<img width="3018" alt="image" src="https://github.com/user-attachments/assets/580d8320-62d3-4627-8640-95c1643dc366" />

### TESTING INSTRUCTIONS
Create a Timeseries or MixedTimeseries chart and use the time grain Quarter.

### ADDITIONAL INFORMATION
<!--- Check any relevant boxes with "x" -->
<!--- HINT: Include "Fixes #nnn" if you are fixing an existing issue -->
- [ ] Has associated issue:
- [ ] Required feature flags:
- [x] Changes UI
- [ ] Includes DB Migration (follow approval process in [SIP-59](https://github.com/apache/superset/issues/13351))
  - [ ] Migration is atomic, supports rollback & is backwards-compatible
  - [ ] Confirm DB migration upgrade and downgrade tested
  - [ ] Runtime estimates and downtime expectations provided
- [ ] Introduces new feature or API
- [ ] Removes existing feature or API


___

## **CodeAnt-AI Description**
**Respect time grain for x-axis and tooltip date formatting in Timeseries and MixedTimeseries charts**

### What Changed
- X-axis labels and tooltip date strings now honor the chart's selected time grain (day, week, month, quarter, year), showing appropriate ranges and concise labels (for example: weekly ranges, "Jan 2021" for months, "Q1 2021" for quarters, "2021" for years).
- Time grain overrides supplied to the chart (for example, via dashboard/extra form overrides) are applied to both x-axis and tooltip formatting.
- The SMART_DATE verbose formatter now produces verbose date strings that match the selected time grain.
- Unit tests added for Timeseries and MixedTimeseries to validate formatting across days, weeks, months, quarters, and years.

### Impact
`✅ Clearer time labels for weekly/monthly/quarterly/yearly charts`
`✅ Consistent tooltip dates with configured time grain`
`✅ Fewer incorrect date displays when dashboards override time grain`
<details>
<summary><strong>💡 Usage Guide</strong></summary>

### Checking Your Pull Request
Every time you make a pull request, our system automatically looks through it. We check for security issues, mistakes in how you're setting up your infrastructure, and common code problems. We do this to make sure your changes are solid and won't cause any trouble later.

### Talking to CodeAnt AI
Got a question or need a hand with something in your pull request? You can easily get in touch with CodeAnt AI right here. Just type the following in a comment on your pull request, and replace "Your question here" with whatever you want to ask:
<pre>
<code>@codeant-ai ask: Your question here</code>
</pre>
This lets you have a chat with CodeAnt AI about your pull request, making it easier to understand and improve your code.

#### Example
<pre>
<code>@codeant-ai ask: Can you suggest a safer alternative to storing this secret?</code>
</pre>

### Preserve Org Learnings with CodeAnt
You can record team preferences so CodeAnt AI applies them in future reviews. Reply directly to the specific CodeAnt AI suggestion (in the same thread) and replace "Your feedback here" with your input:
<pre>
<code>@codeant-ai: Your feedback here</code>
</pre>
This helps CodeAnt AI learn and adapt to your team's coding style and standards.

#### Example
<pre>
<code>@codeant-ai: Do not flag unused imports.</code>
</pre>

### Retrigger review
Ask CodeAnt AI to review the PR again, by typing:
<pre>
<code>@codeant-ai: review</code>
</pre>

### Check Your Repository Health
To analyze the health of your code repository, visit our dashboard at [https://app.codeant.ai](https://app.codeant.ai). This tool helps you identify potential issues and areas for improvement in your codebase, ensuring your repository maintains high standards of code health.

</details>

gerbermichi committed:
Update time formatting and add i18n support for ECharts

rusackas commented:
You need to run the pre-commit hook to pass CI.

Superset uses Git pre-commit hooks courtesy of [pre-commit](https://pre-commit.com/). To install run the following:
```
pip3 install -r requirements/development.txt
pre-commit install
A series of checks will now run when you make a git commit.
```
Alternatively it is possible to run pre-commit by running pre-commit manually:
```
pre-commit run --all-files
```

villebro commented:
@gerbermichi thanks for this PR! A few questions/requests:
1. The PR description is very light on details for all the changes in the PR. Would it be possible to cover all changes in the description to provide more context to reviewers?
2. I see there's a change for both i18n and time grain formatting in this PR - would it be possible to split the PR into two different PRs?
3. I know testing i18n related features can be challenging, but would it be possible to add unit tests for the time grain formatting? Adding a test that failed before but passes after the changes tends to be a great way of showing what the current issue is, how the added/changed logic addresses that, and finally protect against future regressions.

gerbermichi committed:
Add tests for time formatters in xAxis and tooltip

This update introduces unit tests to validate correct time formatting for both xAxis and tooltips in Timeseries and MixedTimeseries charts. The tests ensure proper handling of various time granularities such as days, weeks, months, quarters, and years. This strengthens the reliability and consistency of time display formatting in charts.

gerbermichi commented:
@villebro thank you for your feedback. I removed the i18n part and will add it in a new PR.

I also added tests for the time grain part to ensure that it works as expected.

gerbermichi committed:
respect extra_form_data as well

gerbermichi committed:
use extraFormData instead of extra_form_data

gerbermichi commented:
@villebro can you have a look at this PR?

gerbermichi committed:
Update formatters to handle yearly granularity and verbose date

Replaced quarter-based formatters with yearly ones in test cases to ensure proper handling of year-level granularity. Updated the SMART_DATE format logic to use verbose date formatting for improved clarity in date display.

msyavuz reviewed (commented):
Works fine, just wanted to clarify some things. And you have a failing test

gerbermichi committed:
Update `getSmartDateVerboseFormatter` to accept `timeGrain`

rusackas commented:
Some feedback from Claude:

**Breaking Change Risk**: The convertKeysToCamelCase(formData.extraFormData) could be a breaking change if:

Other components expect snake_case keys from extraFormData
There are existing visualizations that rely on the current key format
The conversion could cause naming conflicts

**Format Override Behavior**: When timeGrain is QUARTER, MONTH, or YEAR, the code completely ignores the user-specified format and uses a default formatter. This might surprise users who expect their custom format to be applied.

**Undefined Format Handling**: The code passes undefined as the format to getTimeFormatter for certain time grains, which needs verification that this is handled properly.

gerbermichi committed:
Remove `convertKeysToCamelCase` and ensure proper handling of `extraFormData` in time formatters

gerbermichi committed:
Simplify form data merging logic in Timeseries and MixedTimeseries transformProps

gerbermichi commented:
> Some feedback from Claude:
> 
> **Breaking Change Risk**: The convertKeysToCamelCase(formData.extraFormData) could be a breaking change if:
> 
> Other components expect snake_case keys from extraFormData There are existing visualizations that rely on the current key format The conversion could cause naming conflicts
> 
> **Format Override Behavior**: When timeGrain is QUARTER, MONTH, or YEAR, the code completely ignores the user-specified format and uses a default formatter. This might surprise users who expect their custom format to be applied.
> 
> **Undefined Format Handling**: The code passes undefined as the format to getTimeFormatter for certain time grains, which needs verification that this is handled properly.

I changed the code so that it does not need to convert the keys. Now I use `formData.extraFormData.time_grain_sqla` directly.

Could you have another look?

gerbermichi committed:
Use `extraFormData` for time grain resolution in Timeseries and MixedTimeseries transformProps
